File: src/cz/vutbr/fit/interlockSim/context/DefaultRailWayNetGrid.java
/* Brno University of Technology
 * Faculty of Information Technology
 * 
 * BSc Thesis	2006/2007
 * 
 * Railway Interlocking Simulator
 * 
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.context;

import java.awt.Point;
import java.util.AbstractSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.Map.Entry;

import cz.vutbr.fit.interlockSim.objects.cells.Cell;
import cz.vutbr.fit.interlockSim.objects.cells.TrackBlockPart;

/**
 * Grid represantation
 */
public class DefaultRailWayNetGrid extends AbstractRailwayNetGrid {
	/**
	 * @see AbstractRailwayNetGrid
	 * @param cols
	 * @param rows
	 */
	public DefaultRailWayNetGrid(int cols, int rows) {
		super(cols, rows);
	}

	private class KeySet extends AbstractSet<Point> {
		private final Set<Entry<Point, Cell>> set;

		KeySet(final Set<Entry<Point, Cell>> set) {
			this.set = set;
		}

		@Override
		public Iterator<Point> iterator() {
			return new Iterator<Point>() {
				private Iterator<Entry<Point, Cell>> i = set.iterator();
				private Entry<Point, Cell> next;
			
				public void remove() {
					if (next == null) throw new IllegalStateException();
					i.remove();
					getReverseTable().remove(next.getValue());
					getReverseTable().values().remove(next.getKey());
				}
			
				public Point next() {
					next = i.next();
					return next.getKey();
				}
			
				public boolean hasNext() {
					return i.hasNext();
				}
			
			};
		}

		@Override
		public int size() {
			return set.size();
		}
	}
	
	/**
	 * @param key
	 * @param cell 
	 * @return previous cell in place
	 */
	public final Cell put(final Point key,final Cell cell) {
		if (getCells().get(key) == cell) return cell;
		getCells().values().remove(cell);
		final Cell prev = getCells().put(key, cell);
		getReverseTable().remove(prev);
		getReverseTable().put(cell, key);
		return prev;
	}

	/**
	 * @param map of point to trackblock part
	 */
	public void putMap(final Map<Point, TrackBlockPart> map) {
		for (Entry<Point, TrackBlockPart> e : map.entrySet()) {
			put(e.getKey(), e.getValue());
		}
	}

	/**
	 * @param newPoint
	 * @return true if point is present 
	 */
	public boolean containsKey(Point newPoint) {
		if (getCells().containsKey(newPoint)) {
			assert getReverseTable().containsValue(newPoint) : newPoint;
			return true;
		}
		assert !getReverseTable().containsValue(newPoint) : newPoint;
		return false;
	}

	/**
	 * Remove cell from grid
	 * @param key
	 */
	public void remove(final Point key) {
		assert key!=null;
		final Cell removed = getCells().remove(key);
		final Point remove2 = getReverseTable().remove(removed);
		assert key.equals(remove2);
	}

	/**
	 * @return true if grid is empty
	 */
	public boolean isEmpty() {
		assert getReverseTable().size() == getCells().size();
		return getCells().size() == 0;
	}

	/**
	 * All cell in grid
	 * @return set of cells
	 */
	public Set<Point> keySet() {
		return new KeySet(getCells().entrySet());
	}
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/context/DefaultRailWayNetGrid.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.context

import cz.vutbr.fit.interlockSim.exceptions.requireValidState
import cz.vutbr.fit.interlockSim.objects.cells.Cell
import cz.vutbr.fit.interlockSim.objects.cells.TrackBlockPart
import cz.vutbr.fit.interlockSim.util.Point
import java.util.AbstractSet

/**
 * Grid represantation
 */
class DefaultRailWayNetGrid(
	cols: Int,
	rows: Int
) : AbstractRailwayNetGrid<Cell>(cols, rows) {
	private inner class KeySet(
		private val set: Set<Map.Entry<Point, Cell>>
	) : AbstractSet<Point>() {
		override fun iterator(): MutableIterator<Point> {
			val delegate: Iterator<Map.Entry<Point, Cell>> = set.iterator()
			return PointIterator(delegate)
		}

		private inner class PointIterator(
			private val delegate: Iterator<Map.Entry<Point, Cell>>
		) : MutableIterator<Point> {
			private var next: Map.Entry<Point, Cell>? = null

			override fun next(): Point {
				next = delegate.next()
				return next!!.key
			}

			override fun hasNext(): Boolean = delegate.hasNext()

			override fun remove() {
				if (delegate is MutableIterator<*>) {
					// Save the cell BEFORE removing from cells map
					// (Map.Entry.getValue() will fail after removal since it looks up the key)
					val cell = next?.value
					// Remove from cells map via delegate
					(delegate as MutableIterator<Map.Entry<Point, Cell>>).remove()
					// Also remove from reverse table to maintain invariant
					if (cell != null) {
						getReverseTable().remove(cell)
					}
				} else {
					throw UnsupportedOperationException("remove")
				}
			}
		}

		override fun removeAll(elements: Collection<Point>): Boolean {
			// Collect cells to remove before modifying the map
			// to avoid ConcurrentModificationException
			val cellsToRemove = mutableListOf<Cell>()
			val pointsToRemove = mutableSetOf<Point>()

			for (point in elements) {
				val cell = getCells()[point]
				if (cell != null) {
					cellsToRemove.add(cell)
					pointsToRemove.add(point)
				}
			}

			// Remove all points from cells map
			for (point in pointsToRemove) {
				getCells().remove(point)
			}

			// Remove all cells from reverse table
			for (cell in cellsToRemove) {
				getReverseTable().remove(cell)
			}

			return pointsToRemove.isNotEmpty()
		}

		override val size: Int
			get() = set.size
	}

	/**
	 * @param key
	 * @param cell
	 * @return previous cell in place
	 */
	@Synchronized
	fun put(
		key: Point,
		cell: Cell
	): Cell? {
		if (getCells().get(key) == cell) return cell
		getCells().values.remove(cell)
		val prev: Cell? = getCells().put(key, cell)
		getReverseTable().remove(prev)
		getReverseTable()[cell] = key
		return prev
	}

	/**
	 * @param map of point to trackblock part
	 */
	@Synchronized
	fun putMap(map: Map<Point, TrackBlockPart>) {
		for ((key, value) in map) {
			put(key, value)
		}
	}

	/**
	 * @param newPoint
	 * @return true if point is present
	 */
	@Synchronized
	override fun containsKey(newPoint: Point): Boolean {
		if (getCells().containsKey(newPoint)) {
			requireValidState(getReverseTable().containsValue(newPoint)) {
				"Inconsistent grid state: point $newPoint in cells but not in reverse table"
			}
			return true
		}
		requireValidState(!getReverseTable().containsValue(newPoint)) {
			"Inconsistent grid state: point $newPoint in reverse table but not in cells: $newPoint"
		}
		return false
	}

	/**
	 * Remove cell from grid
	 * @param key
	 */
	@Synchronized
	fun remove(key: Point) {
		val removed: Cell? = getCells().remove(key)
		val remove2: Point? = getReverseTable().remove(removed)
		requireValidState(key == remove2) {
			"Inconsistent grid state: expected key $key but got $remove2 from reverse table"
		}
	}

	/**
	 * @return true if grid is empty
	 */
	fun isEmpty(): Boolean {
		requireValidState(getReverseTable().isEmpty() == getCells().isEmpty()) {
			"Inconsistent grid state: reverse table and cells have different empty states"
		}
		return getCells().isEmpty()
	}

	/**
	 * All cell in grid
	 * @return set of cells
	 */
	fun keySet(): MutableSet<Point> {
		@Suppress("UNCHECKED_CAST")
		val entries = getCells().entries as Set<Map.Entry<Point, Cell>>
		return KeySet(entries) as MutableSet<Point>
	}
}
